### ISSUE TITLE
Configuration error handling

### ISSUE THREAD

**Issue by spaar** [spaar/besiege-modloader#20]
The handling of possible errors during the loading of the configuration file is currently barely existent.

Ideally, if no file exists, a default one is created (currently already done). If the file is invalid, all possible values are loaded and the remaining ones are set to the default. If this is not possible, reset the whole configuration to default in the event of an error. In either case, report the error in the log.

**Comment by spaar** [comment]
I'm trying to add some basic error handling right now. The problem is that if the loading of the configuration fails, obviously we'd want to print that in the console. However the console doesn't exist yet when the configuration is loaded and depends on it being loaded before being used.
I was thinking of maybe attaching the console before loading the configuration but have the interface (which is the only thing which requires the config) only be enabled after loading a configuration. That way the console would have already registered it's log callback when any errors during loading occur and it could print them.
What do you think, @ZiMMy, @L3tum ? Any better ideas?

**Comment by L3tum** [comment]
It's a good approach!
However, I would think about a variable in the mod loader or somewhere else, a static one, which you can set through the configuration if the loading fails. 
So something like a String, that the error message is applied to, and in the Start() function of Console, you ask if the String != null and then just Debug.LogException it at the end of the function.

**Comment by spaar** [comment]
That might also work, but being able to use Debug.\* directly gives a lot more flexibility.
I currently use Debug.LogWarning if an element is read that's unknown, as that will usually simply indicate that there are either too many elements or at least one default value will be used.
However the serializer won't give me any of the values back if there is a syntax error, so I create a new default config file and Debug.LogError the error.

**Comment by L3tum** [comment]
How are you planning on doing that the Update() function is only called(or the if statement is called) when the config is loaded? A variable inside Keys.cs? Or Console that you set through Keys.cs?

**Comment by spaar** [comment]
That was the part with enabling the interface only after a configuration is available. I created an interfaceEnabled bool in Console and set it to false in the constructor. I also created an EnableInterface method that is called from ModLoader when the configuration loading is done. It sets interfaceEnabled to true. In Update, I simply check if interfaceEnabled is true before calling Keys.

I actually implemented this after writing the first post above and it works quite well. I'm just wondering if there might be a better way to handle this.

**Comment by L3tum** [comment]
It's pretty good. However, I'm not quite sure if I understood you correctly. What do you want to do with it?^^ There will be thrown an exception inside of the Keys.getKey() method, if the Key is unavailable.

**Comment by spaar** [comment]
I know, that's why I don't call Keys.getKey as long as interfaceEnabled is false. And interfaceEnabled is only set to true after the a configuration is available and at that point you are basically guaranteed that every key is available.

**Comment by L3tum** [comment]
Oh, so you want to try to not throw exceptions and just wait for it to load, ok. It's probably better to make it a public static variable in Keys.cs, so that, if needed, every GUI class can access it if it needs it. 
I don't know another method and most of the programs I know doesn't care if something is loaded or not, so I can't look something up.

**Comment by spaar** [comment]
I'm not sure what you mean by wait for it to load. I certainly don't throw any exceptions and I catch any that happen during deserialization. If one is thrown, I print an error message (which is why I need the console already created) and create a new default config instead.
I don't think anything else should need to be loaded before a configuration is available and the console only because it's supposed to also show errors occuring during configuration loading. After the call to Configuration.LoadOrCreateDefault in ModLoader is finished, a configuration is 100% there, containing all keys, even if one could not be loaded. It's just using the default then.

**Comment by spaar** [comment]
So, do you want to look over this or should I just commit it directly?

**Comment by L3tum** [comment]
Oh,ok. I think I misunderstood you a bit, but I couldn't figure out something else, so yeah, commit :)
